examples/platform/silabs/efr32/rs911x/hal/efx_spi.c

```diff
@@ -287,57 +183,60 @@ static void transmitDMA(uint8_t * rx_buf, uint8_t * tx_buf, uint16_t xlen)
  **************************************************************************/
 int16_t rsi_spi_transfer(uint8_t * tx_buf, uint8_t * rx_buf, uint16_t xlen, uint8_t mode)
 {
-    // WFX_RSI_LOG ("SPI: Xfer: tx=%x,rx=%x,len=%d",(uint32_t)tx_buf, (uint32_t)rx_buf, xlen);
-    if (xlen > MIN_XLEN)
+    if (xlen <= MIN_XLEN || (tx_buf == NULL && rx_buf == NULL)) // at least one buffer needs to be provided
     {
-        MY_USART->CMD = USART_CMD_CLEARRX | USART_CMD_CLEARTX;
-        if (xSemaphoreTake(spi_sem, portMAX_DELAY) != pdTRUE)
-        {
-            return RSI_FALSE;
-        }
-        if (tx_buf == NULL)
-        {
-            receiveDMA(rx_buf, xlen);
-        }
-        else
-        {
-            transmitDMA(rx_buf, tx_buf, xlen);
-        }
+        return RSI_ERROR_INVALID_PARAM;
+    }
 
-        /*
-         * receiveDMA() and transmitDMA() are asynchronous
-         * Our application design assumes that this function is synchronous
-         * To make it synchronous, we wait to re-acquire the semaphore before exiting this function
-         * dma_complete_cb() gives back the semaphore when the SPI transfer is done
-         */
-        if (xSemaphoreTake(spi_sem, pdMS_TO_TICKS(RSI_SEM_BLOCK_MIN_TIMER_VALUE_MS)) == pdTRUE)
-        {
-            // Transfer complete
-            // Give back the semaphore before exiting, so that it may be re-acquired
-            // in this function, just before the next transfer
-            xSemaphoreGive(spi_sem);
-        }
-        // Temporary patch
-        // Sometimes the xSemaphoreTake() above is getting stuck indefinitely
-        // As a workaround, if the transfer is not done within RSI_SEM_BLOCK_MIN_TIMER_VALUE_MS
-        // stop and start it again
-        // No need to re-acquire the semaphore since this is the function that acquired it
-        // TODO: Remove this after a permanent solution is found to the problem of the transfer getting stuck
-        else
+    (void) mode; // currently not used;
+    rsi_error_t rsiError = RSI_ERROR_NONE;
+
+    if (xSemaphoreTake(spiTransferLock, portMAX_DELAY) != pdTRUE)
+    {
+        return RSI_ERROR_SPI_BUSY;
+    }
+
+    configASSERT(spiInitiatorTaskHandle == NULL); // No other task should currently be waiting for the dma completion
+    spiInitiatorTaskHandle = xTaskGetCurrentTaskHandle();
+
+    Ecode_t spiError;
+    if (tx_buf == NULL) // Rx operation only
+    {
+        spiError = SPIDRV_MReceive(SPI_HANDLE, rx_buf, xlen, spi_dmaTransfertComplete);
+    }
+    else if (rx_buf == NULL) // Tx operation only
+    {
+        spiError = SPIDRV_MTransmit(SPI_HANDLE, tx_buf, xlen, spi_dmaTransfertComplete);
+    }
+    else // Tx and Rx operation
+    {
+        spiError = SPIDRV_MTransfer(SPI_HANDLE, tx_buf, rx_buf, xlen, spi_dmaTransfertComplete);
+    }
+
```

Comment:
why the SPIDRV doesn't provide an helper for this within the SDK

Reply:
you'd have to ask the platform team that implemented the spidrv. I don't have an answer